src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,367 @@
+            createSession: createSessionUrl,
+            uploadPart: null,
+            listParts: null,
+            commit: null,
+            abort: null,
+            logEvent: null
+        };
+        // Set a default logEvent endpoint so that we can report failures from createSession.
+        // This value will be overwritten on successful response from createSession.
+        // This assumes that the createSessionUrl already contains the upload session id.
+        this.sessionEndpoints.logEvent = `${createSessionUrl.replace(/files\/\d+/, 'files')}/log`;
+        // Explicitly cast these two to string because sometimes they get passed as int, and
+        // sending folder_id as a number instead of a string in the JSON body of createSession
+        // results in an API error.
+        this.destinationFolder = destinationFolder;
+        this.destinationFile = destinationFile;
+        this.config = config || DEFAULT_MULTIPUT_CONFIG;
+        this.fileSha1 = null;
+        this.totalUploadedBytes = 0;
+        this.partsNotStarted = 0; // # of parts yet to be processed
+        this.partsDigestComputing = 0; // # of parts sent to the digest worker
+        this.partsDigestReady = 0; // # of parts with digest finished that are waiting to be uploaded.
+        this.partsUploading = 0; // # of parts with upload requests currently inflight
+        this.partsUploaded = 0; // # of parts successfully uploaded
+        this.firstUnuploadedPartIndex = 0; // Index of first part that hasn't been uploaded yet.
+        this.onProgress = null;
+        this.onCompleted = null;
+        this.onFailure = null;
+        this.createSessionNumRetriesPerformed = 0;
+        this.partSize = 0;
+        this.parts = [];
+        this.commitRetryCount = 0;
+        this.ended = false;
+        this.worker = createWorker();
+        this.clientId = null;
+    }
+
+    /**
+     * Creates upload session. If a file ID is supplied, use the Chunked Upload File Version
+     * API to replace the file.
+     *
+     * @private
+     * @param {Object} - Request options
+     * @param {boolean} [options.url] - Upload URL to use
+     * @param {string} [options.fileId] - ID of file to replace
+     * @param {string} [options.fileName] - New name for file - required for new file upload
+     * @return {void}
+     */
+    createUploadSession = async (): Promise<> => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        // Set up post body
+        const postData: StringAnyMap = {
+            file_size: this.file.size,
+            file_name: this.file.name
+        };
+
+        if (this.destinationFolder) {
+            postData.folder_id = this.destinationFolder;
+        }
+
+        try {
+            const data = await this.xhr.post({ url: this.sessionEndpoints.createSessionUrl, data: postData });
+            this.uploadSessionSuccessHandler(data);
+        } catch (error) {
+            this.uploadSessionErrorHandler(error);
```

Comment:
Do non-200 status codes cause the error handler to be triggered?

Reply:
yes they should

Reply:
Will the error handler receive the status code? I believe we have logic in our error handler to recover from a 409 Conflict (which can occur on a retry, where the session was actually created the first time around)

Reply:
Yes, I believe it'll be in `error.status`